### PR TITLE
feature: oneMKL migration

### PR DESCRIPTION
**Pull request by Alexandr-Solovev**
### Description:
**Feature: Migration from FPKs to oneMKL**

#### Summary:
This PR updates the oneDAL codebase to ensure compatibility with oneMKL. It includes various fixes and modifications across multiple files to address issues related to the integration with the new libraries.

#### Key Changes:
1. **Updating Public CI Environment:**
   - Since MKL FPKs are no longer used, publicly available oneMKL packages have been added to the CI environment.

2. **Removing macOS Public CI Step:**
   - As oneMKL does not support macOS, the macOS step has been removed from the CI pipeline.

3. **Bazel Changes:**
   - Instead of using two different Bazel libraries (mklcpu and mklgpu), a unified MKL library is now utilized.

4. **Service Functions:**
   - The first set of FPK service functions has been replaced with public oneMKL functions, and the remaining ones have been replaced with standard C++ library functions.

5. **Naming Updates:**
   - References to FPK have been removed or replaced with MKL.

6. **Macros Usage and CPU Instructions Updates:**
   - As oneMKL does not allow switching between CPU instruction sets, all `DAAL_MKL_CALL` macros have been replaced with `fname_fargs` macros to call the equivalent functions in oneMKL.

7. **Sparse API Updates:**
   - The deprecated spBLAS stack from MKL has been replaced with the new oneMKL Inspector-Executor API.

8. **PCA Algorithm Refactoring:**
   - With the removal of the micromkl library, the PCA algorithm has been migrated to use GPU functions (syevd, gesvd).

9. **Makefile Updates:**
   - Library names and paths have been updated to support oneMKL.

#### Known Issues/Looking for Help:

- [x] 1. **Syevd and Gesvd Tests:**
   - Assistance needed with these tests.

- [x] 2. **Public Build for Windows (WindowsMakeVC Step):**

   - Guidance required to obtain a public build for this step.

- [x] 3. **Removing Unnecessary CPU Instructions and Macros:**

   - Help needed with this cleanup.

- [x] 4. **Fixing Path Substitution for Windows:**

   - Need to correct the substitution of `\`, `/`, and `$` for proper Windows path handling.

### CODE REVIEW

**Comment by icfaust** [comment]
@Alexandr-Solovev You'll need to work on mircomkl integration in the CI systems, and please include an informative description (or is this still supposed to be in draft?)

File: cpp/daal/src/externals/service_blas_mkl.h
Comment (olegkkruglov) [inline cpp/daal/src/externals/service_blas_mkl.h:42]: is it possible to avoid this macros redefinition? what is the point of it?

File: cpp/daal/src/externals/service_math_mkl.h
Comment (olegkkruglov) [inline cpp/daal/src/externals/service_math_mkl.h:36]: why is f_suff arg necessary here?

File: cpp/daal/src/externals/service_stat_mkl.h
Comment (olegkkruglov) [inline cpp/daal/src/externals/service_stat_mkl.h:167]: why is this number hardcoded?

File: cpp/daal/src/externals/service_stat_mkl.h
Comment (olegkkruglov) [inline cpp/daal/src/externals/service_stat_mkl.h:151]: what do 0 and 1 mean here?

File: cpp/daal/src/externals/service_stat_mkl.h
Comment (olegkkruglov) [inline cpp/daal/src/externals/service_stat_mkl.h:149]: do I understand correctly that parallel for is replaced by usual one here? if so, does it affect performance?

File: cpp/daal/src/externals/service_stat_mkl.h
Comment (olegkkruglov) [inline cpp/daal/src/externals/service_stat_mkl.h:207]: since threading structure is deleted and it was the only place its elements are used, does it safe to completely delete them?

File: cpp/daal/src/externals/service_stat_rng_mkl.h
Comment (olegkkruglov) [inline cpp/daal/src/externals/service_stat_rng_mkl.h:34]: what is the point of this change?

File: cpp/daal/src/externals/service_stat_rng_mkl.h
Comment (olegkkruglov) [inline cpp/daal/src/externals/service_stat_rng_mkl.h:80]: should it be deleted?

File: cpp/daal/src/services/library_version_info.cpp
Comment (olegkkruglov) [inline cpp/daal/src/services/library_version_info.cpp:53]: probably the comment must be deleted as well

File: cpp/oneapi/dal/algo/pca/backend/gpu/finalize_train_kernel_cov_impl_dpc.cpp
Comment (olegkkruglov) [inline cpp/oneapi/dal/algo/pca/backend/gpu/finalize_train_kernel_cov_impl_dpc.cpp:106]: I guess event dependencies must be reviewed here to avoid unnecessary wait_and_throw

File: cpp/oneapi/dal/algo/pca/backend/gpu/train_kernel_cov_impl_dpc.cpp
Comment (olegkkruglov) [inline cpp/oneapi/dal/algo/pca/backend/gpu/train_kernel_cov_impl_dpc.cpp:125]: https://github.com/oneapi-src/oneDAL/pull/2865/files#r1733446937

File: cpp/oneapi/dal/backend/primitives/lapack/test/eigen.cpp
Comment (olegkkruglov) [inline cpp/oneapi/dal/backend/primitives/lapack/test/eigen.cpp:14]: why was this test deleted?

File: dev/make/deps.mkl.mk
Comment (olegkkruglov) [inline dev/make/deps.mkl.mk:21]: please update naming to avoid "FPK" usage

File: makefile
Comment (olegkkruglov) [inline makefile:480]: should ILP64 flag be backend dependent? what if non-MKL BACKEND_CONFIG is used?

File: dev/make/deps.mkl.mk
Comment (olegkkruglov) [inline dev/make/deps.mkl.mk:35]:
```suggestion
daaldep.math_backend_oneapi.incdir := $(subst \,/,$(MKLFPKDIR.include) subst \,/,$(MKLGPUFPKDIR.include))
```

Comment (Alexandr-Solovev) [reply dev/make/deps.mkl.mk]: Its not necessary to subst in this variable at all with new updates

File: makefile
Comment (Alexandr-Solovev) [reply makefile:528]: The oneapi backend include SYCL:
backend.incdir=cpu mkl
backend_oneapi.incdir=cpu+gpu mkl

File: .ci/pipeline/ci.yml
Comment (Alexsandruss) [inline .ci/pipeline/ci.yml:671]:
```suggestion
    displayName: 'Install oneAPI Base Toolkit'
```

File: .ci/pipeline/ci.yml
Comment (Alexsandruss) [inline .ci/pipeline/ci.yml:683]: Why GPU FPK is still mentioned?

File: dev/make/compiler_definitions/dpcpp.mk
Comment (Alexsandruss) [inline dev/make/compiler_definitions/dpcpp.mk:35]: Do not forget to return or remove all commented parts.

File: .ci/pipeline/ci.yml
Comment (Alexandr-Solovev) [reply .ci/pipeline/ci.yml:683]: Its necessary for choosing correct layout of oneTBB lib, we still have in makefile this flag. Flag decides is it lib/intel64 or lib/ layout

File: cpp/daal/src/algorithms/naivebayes/naivebayes_predict_kernel.h
Comment (Vika-F) [inline cpp/daal/src/algorithms/naivebayes/naivebayes_predict_kernel.h:25]:
```suggestion
#ifndef __NAIVEBAYES_PREDICT_KERNEL_H__
#define __NAIVEBAYES_PREDICT_KERNEL_H__
```

File: cpp/daal/src/externals/service_blas_mkl.h
Comment (Vika-F) [inline cpp/daal/src/externals/service_blas_mkl.h:40]: Looks like `__DAAL_CONCAT4` and similar macros can be removed also.

File: cpp/daal/src/externals/service_math_mkl.h
Comment (Vika-F) [inline cpp/daal/src/externals/service_math_mkl.h:109]: MKL math functions support various accuracy modes.
Previously the `mode` parameter was provided as an argument to a function (`HAynn`) and now it is not. This might lead to unexpected changes in the functions results.
Let's provide the mode to the functions explicitly like it was done previously.

To do this you need to call `vmdPowx` instead `vdPowx`, for example.
and provide following set of flags as the additional argument at the end of the arguments list: `VML_HA | VML_FTZDAZ_CURRENT | VML_ERRMODE_DEFAULT`.
See https://www.intel.com/content/www/us/en/docs/onemkl/developer-reference-c/2023-2/v-powx.html

File: cpp/daal/src/externals/service_service_mkl.h
Comment (Vika-F) [inline cpp/daal/src/externals/service_service_mkl.h:49]: `memcpy_s` and other safe functions are available in C++11. Can you try to use those?

File: cpp/daal/src/externals/service_service_mkl.h
Comment (Vika-F) [inline cpp/daal/src/externals/service_service_mkl.h:66]: The implementation should be similar to what is done in this test
https://github.com/oneapi-src/oneDAL/blob/main/cpp/oneapi/dal/test/global_context.cpp#L97
Need to check the bits [23:16] of ebx register after calling `cpuid(0x1, registers);`. I can help with the actual implementation.

File: dev/bazel/deps/mkl.bzl
Comment (samir-nasibli) [inline dev/bazel/deps/mkl.bzl:27]: Is this required?

File: dev/make/deps.mkl.mk
Comment (samir-nasibli) [inline dev/make/deps.mkl.mk:74]: Why this is extended for threading?

**Comment by samir-nasibli** [comment]
@Alexandr-Solovev please mark this as a DRAFT, if it is still under the development.

File: cpp/daal/src/externals/service_service_mkl.h
Comment (Alexandr-Solovev) [reply cpp/daal/src/externals/service_service_mkl.h:50]: I was trying but it returns undefined_references to these functions even with different headers. Looks like can be added later

File: dev/bazel/deps/mkl.bzl
Comment (Alexandr-Solovev) [reply dev/bazel/deps/mkl.bzl:26]: yes, previously we had 2 different mkl deps in bazel, and it was split as seq and thread versions. Now we agreed that we will use thread mkl in all scenarios, so, I replaced "lib/libmkl_sequential.a", with "lib/libmkl_tbb_thread.a",

File: .ci/scripts/install_windows.bat
Comment (Vika-F) [inline .ci/scripts/install_windows.bat:3]:
```suggestion
rem Copyright contributors to the oneDAL project
```

File: WORKSPACE
Comment (Vika-F) [inline WORKSPACE:75]: `TODO` makes it more searchable.
```suggestion
        # TODO: when the issue with binutils will be solved, replace 2023.0 to 2024.2
```

File: cpp/daal/src/externals/service_math_mkl.h
Comment (Vika-F) [inline cpp/daal/src/externals/service_math_mkl.h:106]: The precision was dropped a bit for all functions with this change. It was VML_HA (i.e. "high accuracy"), and now it is VML_LA ("low accuracy").
Maybe not a big deal for machine learning algorithms. But maybe it contributed to some test failures in sklex.
Need to return to this after RC.
```suggestion
        __DAAL_MKLFN_CALL_MATH(vmdPowx, ((int)n, in, in1, out, (VML_HA | VML_FTZDAZ_ON | VML_ERRMODE_STDERR)));
```

File: cpp/oneapi/dal/algo/pca/backend/gpu/misc.hpp
Comment (Vika-F) [inline cpp/oneapi/dal/algo/pca/backend/gpu/misc.hpp:69]: Strange comments here.

**Comment by Alexandr-Solovev** [comment]
Link to the job with oneMKL: http://intel-ci.intel.com/ef7b39b8-64f0-f1ec-91a0-a4bf010d0e2d

File: WORKSPACE
Comment (samir-nasibli) [inline WORKSPACE:75]: Please cover this with a ticket

File: WORKSPACE
Comment (samir-nasibli) [inline WORKSPACE:89]: Not required?

File: cpp/daal/src/externals/service_service_mkl.h
Comment (samir-nasibli) [inline cpp/daal/src/externals/service_service_mkl.h:53]: Please add some ticket for this kind of refactoring

File: cpp/daal/src/externals/service_service_mkl.h
Comment (samir-nasibli) [inline cpp/daal/src/externals/service_service_mkl.h:99]: Ticket

File: cpp/daal/src/externals/service_stat_mkl.h
Comment (samir-nasibli) [reply cpp/daal/src/externals/service_stat_mkl.h:207]: I guess this far somehow discussed, make sense for reply comment with the answer before marking it resolved

File: cpp/daal/src/externals/service_thread_declar_mkl.cpp
Comment (samir-nasibli) [inline cpp/daal/src/externals/service_thread_declar_mkl.cpp:25]: Ticket

File: cpp/oneapi/dal/algo/pca/backend/gpu/train_kernel_cov_impl_dpc.cpp
Comment (samir-nasibli) [inline cpp/oneapi/dal/algo/pca/backend/gpu/train_kernel_cov_impl_dpc.cpp:73]: Just for my understanding: Why we adding these blocking waits, making this async?

**Comment by samir-nasibli** [comment]
Question: do we need to update `scikit-learn-intelex` deps in case of these migration?

**Review by samir-nasibli — COMMENTED** [review]
Huge work done. Thank you @asolovyo !
I am sure we should have follow up tickets for all TODOs.
Provided just minor comments.

For merging, I recommend getting approval from other reviewers as well.

File: .ci/env/apt.sh
Comment (Alexandr-Solovev) [reply .ci/env/apt.sh:38]: I guess it wasn't used before, since I added it, it starts to fail without -y flag

**Comment by samir-nasibli** [comment]
https://github.com/intel/scikit-learn-intelex/pull/2066 is merged.
@Alexandr-Solovev you can launch the CI for the scikit-learn-intelex main.

File: WORKSPACE
Comment (Alexandr-Solovev) [reply WORKSPACE:89]: 2024.2 version is built with binutils 2.37, currently we use in our infra 2.30, and it calls a lot of warnings, it will be updated after merging all current prs

File: cpp/oneapi/dal/algo/pca/backend/gpu/train_kernel_cov_impl_dpc.cpp
Comment (Alexandr-Solovev) [reply cpp/oneapi/dal/algo/pca/backend/gpu/train_kernel_cov_impl_dpc.cpp:73]: we observe strange events control work when we use syevd function from oneMKL, these wait_and_throw functions are a workaround

**Review by Vika-F — APPROVED** [review]
Great work! The changes look good to me.

**Pull request merged**
